uasc/secure_channel.go

```diff
@@ -114,258 +94,202 @@ func NewSecureChannel(endpoint string, c *uacp.Conn, cfg *Config) (*SecureChanne
+					return resp
+				}
+				s.chunksMu.Unlock()
+				continue
+			}
 
-	// encode the message
-	return s.newMessage(req, typeID), nil
-}
+			// merge chunks
+			all := append(s.chunks[reqID], chunk)
+			delete(s.chunks, reqID)
 
-// SendResponse sends a service response.
-// todo(fs): this method is most likely needed for the server and we haven't tested it yet.
-// todo(fs): it exists to implement the handleOpenSecureChannelRequest() method during the
-// todo(fs): refactor to remove the reflect code. It will likely change.
-func (s *SecureChannel) SendResponse(req ua.Response) error {
-	typeID := ua.ServiceTypeID(req)
-	if typeID == 0 {
-		return errors.Errorf("unknown service %T. Did you call register?", req)
-	}
+			s.chunksMu.Unlock()
 
-	// encode the message
-	m := s.newMessage(req, typeID)
-	reqid := m.SequenceHeader.RequestID
-	b, err := m.Encode()
-	if err != nil {
-		return err
-	}
+			b, err := mergeChunks(all)
+			if err != nil {
+				resp.Err = err
+				return resp
+			}
 
-	// encrypt the message prior to sending it
-	// if SecurityMode == None, this returns the byte stream untouched
-	b, err = s.signAndEncrypt(m, b)
-	if err != nil {
-		return err
-	}
+			if uint32(len(b)) > s.c.MaxMessageSize() {
+				resp.Err = errors.Errorf("message too large: %d > %d", uint32(len(b)), s.c.MaxMessageSize())
+				return resp
+			}
 
-	// send the message
-	if _, err := s.c.Write(b); err != nil {
-		return err
-	}
-	debug.Printf("uasc %d/%d: send %T with %d bytes", s.c.ID(), reqid, req, len(b))
+			// since we are not decoding the ResponseHeader separately
+			// we need to drop every message that has an error since we
+			// cannot get to the RequestHandle in the ResponseHeader.
+			// To fix this we must a) decode the ResponseHeader separately
+			// and subsequently remove it and the TypeID from all service
+			// structs and tests. We also need to add a deadline to all
+			// handlers and check them periodically to time them out.
+			_, svc, err := ua.DecodeService(b)
+			if err != nil {
+				resp.Err = err
+				return resp
+			}
 
-	return nil
+			resp.V = svc
+
+			// If the service status is not OK then bubble
+			// that error up to the caller.
+			if r, ok := svc.(ua.Response); ok {
+				if status := r.Header().ServiceResult; status != ua.StatusOK {
+					resp.Err = status
+					return resp
+				}
+			}
+
+			return resp
+		}
+	}
 }
 
 func (s *SecureChannel) readChunk() (*MessageChunk, error) {
 	// read a full message from the underlying conn.
 	b, err := s.c.Receive()
-	if err == io.EOF || s.hasState(secureChannelClosed) {
+	if err == io.EOF || len(b) == 0 { // || s.hasState(secureChannelClosed) {
 		return nil, io.EOF
 	}
-	if errf, ok := err.(*uacp.Error); ok {
-		return nil, errf
-	}
+
 	if err != nil {
 		return nil, errors.Errorf("sechan: read header failed: %s %#v", err, err)
 	}
 
-	const hdrlen = 12
+	const hdrlen = 12 // TODO: move to pkg level const
```

Comment:
I think the reason I put this here is because I wanted a constant to have a name for a magic number but that number was only ever going to be used in this tiny context.

Reply:
IIRC I came across the same value in another file/func.